File: app/src/main/java/com/example/doan/Activities/ProductDetailActivity.java
package com.example.doan.Activities;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;
import com.example.doan.Models.AddToCartRequest;
import com.example.doan.Models.ApiResponse;
import com.example.doan.Models.CreateOrderRequest;
import com.example.doan.Models.Drink;
import com.example.doan.Models.DrinkSize;
import com.example.doan.Models.DrinkTopping;
import com.example.doan.Models.Order;
import com.example.doan.Models.OrderItemRequest;
import com.example.doan.Models.Store;
import com.example.doan.R;
import com.example.doan.Network.RetrofitClient;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ProductDetailActivity extends AppCompatActivity {
    private static final String TAG = "ProductDetailActivity";

    private ImageView ivProductImage;
    private TextView tvProductName, tvProductPrice, tvProductDescription;
    private TextView tvQuantity, tvTotalPrice;
    private com.google.android.material.button.MaterialButton btnDecrease, btnIncrease;
    private com.google.android.material.button.MaterialButton btnAddToCart, btnConfirmOrder;
    private Spinner spinnerSize, spinnerBranch;
    private LinearLayout layoutToppings;
    private ImageButton btnBack;

    private Drink product;
    private int quantity = 1;
    private DrinkSize selectedSize;
    private final Set<DrinkTopping> selectedToppings = new HashSet<>();
    private List<Store> storeList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product_detail);

        initViews();
        getProductDetails();
        loadStores();
        setupListeners();
    }

    private int getLoggedInUserId() {
        // Thử SessionManager trước
        com.example.doan.Utils.SessionManager sessionManager = new com.example.doan.Utils.SessionManager(this);
        if (sessionManager.isLoggedIn()) {
            return sessionManager.getUserId();
        }
        
        // Fallback: Thử SharedPreferences cũ
        SharedPreferences sharedPreferences = getSharedPreferences("UserPrefs", MODE_PRIVATE);
        return sharedPreferences.getInt("userId", -1);
    }

    private void initViews() {
        ivProductImage = findViewById(R.id.ivProductImage);
        tvProductName = findViewById(R.id.tvProductName);
        tvProductPrice = findViewById(R.id.tvProductPrice);
        tvProductDescription = findViewById(R.id.tvProductDescription);
        tvQuantity = findViewById(R.id.tvQuantity);
        tvTotalPrice = findViewById(R.id.tvTotalPrice);
        btnDecrease = findViewById(R.id.btnDecrease);
        btnIncrease = findViewById(R.id.btnIncrease);
        btnAddToCart = findViewById(R.id.btnAddToCart);
        btnConfirmOrder = findViewById(R.id.btnConfirmOrder);
        spinnerSize = findViewById(R.id.spinnerSize);
        spinnerBranch = findViewById(R.id.spinnerBranch);
        layoutToppings = findViewById(R.id.layoutToppings);

        btnBack = findViewById(R.id.btnBack);
        btnBack.setOnClickListener(v -> finish());
    }

    private void getProductDetails() {
        Intent intent = getIntent();
        if (intent != null && intent.hasExtra("product")) {
            String productJson = intent.getStringExtra("product");
            Log.d(TAG, "Product JSON: " + productJson);
            try {
                // Try parsing as Drink first
                product = new Gson().fromJson(productJson, Drink.class);
                if (product != null) {
                    Log.d(TAG, "Product parsed successfully: " + product.getName());
                    Log.d(TAG, "Product ID: " + product.getId());
                    Log.d(TAG, "Product basePrice: " + product.getBasePrice());
                    Log.d(TAG, "Product imageUrl: " + product.getImageUrl());
                    displayProductDetails();
                } else {
                    Log.e(TAG, "Product is null after parsing");
                    Toast.makeText(this, "Không thể tải chi tiết sản phẩm", Toast.LENGTH_SHORT).show();
                    finish();
                }
            } catch (Exception e) {
                Log.e(TAG, "Error parsing product: " + e.getMessage());
                e.printStackTrace();
                Toast.makeText(this, "Lỗi tải sản phẩm: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                finish();
            }
        } else {
            Log.e(TAG, "No product extra in intent");
            Toast.makeText(this, "Không có thông tin sản phẩm", Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    private void displayProductDetails() {
        try {
            Log.d(TAG, "Displaying product details");
            
            if (product.getName() != null) {
                tvProductName.setText(product.getName());
            } else {
                tvProductName.setText("Tên sản phẩm");
            }
            
            if (product.getDescription() != null) {
                tvProductDescription.setText(product.getDescription());
            } else {
                tvProductDescription.setText("Không có mô tả");
            }
            
            tvProductPrice.setText(String.format(Locale.getDefault(), "%,.0f VNĐ", product.getBasePrice()));

            if (product.getImageUrl() != null && !product.getImageUrl().isEmpty()) {
                Log.d(TAG, "Loading image: " + product.getImageUrl());
                Glide.with(this)
                    .load(product.getImageUrl())
                    .placeholder(R.drawable.ic_image_placeholder)
                    .error(R.drawable.ic_image_placeholder)
                    .into(ivProductImage);
            }

            setupSizeSpinner();
            setupToppingCheckboxes();
            updateTotalPrice();
            
            Log.d(TAG, "Product details displayed successfully");
        } catch (Exception e) {
            Log.e(TAG, "Error displaying product details: " + e.getMessage());
            e.printStackTrace();
            Toast.makeText(this, "Lỗi hiển thị sản phẩm: " + e.getMessage(), Toast.LENGTH_SHORT).show();
            finish();
        }
    }

    private void loadStores() {
        RetrofitClient.getInstance(this).getApiService().getStores().enqueue(new Callback<ApiResponse<List<Store>>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResponse<List<Store>>> call, @NonNull Response<ApiResponse<List<Store>>> response) {
                if (response.isSuccessful() && response.body() != null && response.body().isSuccess()) {
                    storeList = response.body().getData();
                    if (storeList != null && !storeList.isEmpty()) {
                        List<String> storeNames = new ArrayList<>();
                        for (Store store : storeList) {
                            storeNames.add(store.getStoreName());
                        }
                        ArrayAdapter<String> storeAdapter = new ArrayAdapter<>(ProductDetailActivity.this, android.R.layout.simple_spinner_item, storeNames);
                        storeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                        spinnerBranch.setAdapter(storeAdapter);
                    }
                } else {
                    Toast.makeText(ProductDetailActivity.this, "Không thể tải danh sách cửa hàng", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResponse<List<Store>>> call, @NonNull Throwable t) {
                Log.e(TAG, "Lỗi tải danh sách cửa hàng: " + t.getMessage());
                Toast.makeText(ProductDetailActivity.this, "Lỗi kết nối", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void setupListeners() {
        btnDecrease.setOnClickListener(v -> {
            if (quantity > 1) {
                quantity--;
                tvQuantity.setText(String.valueOf(quantity));
                updateTotalPrice();
            }
        });

        btnIncrease.setOnClickListener(v -> {
            quantity++;
            tvQuantity.setText(String.valueOf(quantity));
            updateTotalPrice();
        });

        btnConfirmOrder.setOnClickListener(v -> placeOrder());
        
        btnAddToCart.setOnClickListener(v -> addToCart());
    }

    private void setupSizeSpinner() {
        if (product.getSizes() == null || product.getSizes().isEmpty()) {
            spinnerSize.setVisibility(View.GONE);
            // Không set selectedSize, sẽ gửi null lên backend
            selectedSize = null;
            return;
        }
        
        List<String> sizeOptions = new ArrayList<>();
        for (DrinkSize size : product.getSizes()) {
            String option = size.getSizeName();
            if (size.getExtraPrice() > 0) {
                option += " (+" + String.format(Locale.getDefault(), "%,.0f VNĐ", size.getExtraPrice()) + ")";
            }
            sizeOptions.add(option);
        }
        
        ArrayAdapter<String> sizeAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, sizeOptions);
        sizeAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinnerSize.setAdapter(sizeAdapter);
        
        // Set default selection
        selectedSize = product.getSizes().get(0);
        
        spinnerSize.setOnItemSelectedListener(new android.widget.AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(android.widget.AdapterView<?> parent, View view, int position, long id) {
                selectedSize = product.getSizes().get(position);
                updateTotalPrice();
                Log.d(TAG, "Size selected: " + selectedSize.getSizeName() + ", Extra: " + selectedSize.getExtraPrice());
            }
            
            @Override
            public void onNothingSelected(android.widget.AdapterView<?> parent) {}
        });
    }
    
    private void setupToppingCheckboxes() {
        if (product.getToppings() == null || product.getToppings().isEmpty()) {
            layoutToppings.setVisibility(View.GONE);
            return;
        }
        
        layoutToppings.removeAllViews();
        
        for (DrinkTopping topping : product.getToppings()) {
            if (!topping.isActive()) continue;
            
            android.widget.CheckBox checkBox = new android.widget.CheckBox(this);
            checkBox.setText(topping.getToppingName() + " (+" + String.format(Locale.getDefault(), "%,.0f VNĐ", topping.getPrice()) + ")");
            checkBox.setTextSize(14);
            checkBox.setPadding(8, 8, 8, 8);
            
            checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (isChecked) {
                    selectedToppings.add(topping);
                    Log.d(TAG, "Topping added: " + topping.getToppingName());
                } else {
                    selectedToppings.remove(topping);
                    Log.d(TAG, "Topping removed: " + topping.getToppingName());
                }
                updateTotalPrice();
            });
            
            layoutToppings.addView(checkBox);
        }
    }
    
    private void updateTotalPrice() {
        double basePrice = product.getBasePrice();
        double sizeExtra = selectedSize != null ? selectedSize.getExtraPrice() : 0;
        double toppingTotal = 0;
        
        for (DrinkTopping topping : selectedToppings) {
            toppingTotal += topping.getPrice();
        }
        
        double itemPrice = basePrice + sizeExtra + toppingTotal;
        double total = itemPrice * quantity;
        
        tvTotalPrice.setText(String.format(Locale.getDefault(), "%,.0f VNĐ", total));
        
        Log.d(TAG, "Price breakdown - Base: " + basePrice + ", Size: " + sizeExtra + ", Toppings: " + toppingTotal + ", Total: " + total);
    }

    private void addToCart() {
        int userId = getLoggedInUserId();
        if (userId == -1) {
            Toast.makeText(this, "Vui lòng đăng nhập để thêm vào giỏ hàng", Toast.LENGTH_SHORT).show();
            return;
        }
        
        // Tạo danh sách topping IDs
        List<Long> toppingIds = new ArrayList<>();
        for (DrinkTopping topping : selectedToppings) {
            toppingIds.add((long) topping.getId());
        }
        
        // Tạo note với thông tin topping
        StringBuilder note = new StringBuilder();
        if (!selectedToppings.isEmpty()) {
            note.append("Topping: ");
            for (int i = 0; i < selectedToppings.size(); i++) {
                if (i > 0) note.append(", ");
                note.append(selectedToppings.toArray(new DrinkTopping[0])[i].getToppingName());
            }
        }
        
        // Nếu không có size, gửi null
        Long sizeId = selectedSize != null ? (long) selectedSize.getId() : null;
        
        com.example.doan.Models.AddToCartRequest request = new com.example.doan.Models.AddToCartRequest(
            (long) product.getId(),
            sizeId,
            quantity,
            toppingIds.isEmpty() ? null : toppingIds,
            note.toString()
        );
        
        btnAddToCart.setEnabled(false);
        btnAddToCart.setText("Đang thêm...");
        
        RetrofitClient.getInstance(this).getApiService().addToCart(userId, request)
            .enqueue(new Callback<ApiResponse<com.example.doan.Models.Cart>>() {
                @Override
                public void onResponse(@NonNull Call<ApiResponse<com.example.doan.Models.Cart>> call, 
                                     @NonNull Response<ApiResponse<com.example.doan.Models.Cart>> response) {
                    btnAddToCart.setEnabled(true);
                    btnAddToCart.setText("THÊM VÀO GIỎ");
                    
                    if (response.isSuccessful() && response.body() != null && response.body().isSuccess()) {
                        // Hiển thị dialog hỏi user
                        new androidx.appcompat.app.AlertDialog.Builder(ProductDetailActivity.this)
                            .setTitle("Thêm vào giỏ hàng thành công!")
                            .setMessage("Bạn có muốn xem giỏ hàng ngay không?\n\n⚠️ Lưu ý: Giỏ hàng sẽ tự động xóa sau 5 phút nếu không đặt hàng.")
                            .setPositiveButton("Xem giỏ hàng", (dialog, which) -> {
                                startActivity(new Intent(ProductDetailActivity.this, CartActivity.class));
                            })
                            .setNegativeButton("Tiếp tục mua", (dialog, which) -> {
                                dialog.dismiss();
                            })
                            .show();
                    } else {
                        String message = response.body() != null ? response.body().getMessage() : "Lỗi thêm vào giỏ hàng";
                        Toast.makeText(ProductDetailActivity.this, message, Toast.LENGTH_SHORT).show();
                    }
                }
                
                @Override
                public void onFailure(@NonNull Call<ApiResponse<com.example.doan.Models.Cart>> call, @NonNull Throwable t) {
                    btnAddToCart.setEnabled(true);
                    btnAddToCart.setText("THÊM VÀO GIỎ");
                    Log.e(TAG, "Lỗi thêm vào giỏ hàng: " + t.getMessage());
                    Toast.makeText(ProductDetailActivity.this, 
                        "Không thể kết nối đến máy chủ", Toast.LENGTH_SHORT).show();
                }
            });
    }

    private void placeOrder() {
        if (storeList.isEmpty()) {
            Toast.makeText(this, "Đang tải danh sách cửa hàng, vui lòng đợi...", Toast.LENGTH_SHORT).show();
            return;
        }

        int selectedStoreIndex = spinnerBranch.getSelectedItemPosition();
        if (selectedStoreIndex == -1) {
            Toast.makeText(this, "Vui lòng chọn cửa hàng", Toast.LENGTH_SHORT).show();
            return;
        }
        
        long storeId = storeList.get(selectedStoreIndex).getId();
        
        int userId = getLoggedInUserId();
        if (userId == -1) {
            Toast.makeText(this, "Vui lòng đăng nhập để đặt hàng", Toast.LENGTH_SHORT).show();
            return;
        }
        
        // Tạo danh sách topping IDs
        List<Long> toppingIds = new ArrayList<>();
        for (DrinkTopping topping : selectedToppings) {
            toppingIds.add((long) topping.getId());
        }
        
        // Tạo note với thông tin topping
        StringBuilder note = new StringBuilder();
        if (!selectedToppings.isEmpty()) {
            note.append("Topping: ");
            for (int i = 0; i < selectedToppings.size(); i++) {
                if (i > 0) note.append(", ");
                note.append(selectedToppings.toArray(new DrinkTopping[0])[i].getToppingName());
            }
        }
        
        // Nếu không có size, dùng "M" làm mặc định
        String sizeName = selectedSize != null ? selectedSize.getSizeName() : "M";
        
        OrderItemRequest item = new OrderItemRequest(
            (long) product.getId(), 
            sizeName, 
            quantity, 
            note.toString(),
            toppingIds.isEmpty() ? null : toppingIds
        );

        CreateOrderRequest orderRequest = new CreateOrderRequest(storeId, "PICKUP", "Tại cửa hàng", "COD", Collections.singletonList(item));

        btnConfirmOrder.setEnabled(false);
        btnConfirmOrder.setText("Đang xử lý...");

        RetrofitClient.getInstance(this).getApiService().createOrder(orderRequest).enqueue(new Callback<ApiResponse<Order>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResponse<Order>> call, @NonNull Response<ApiResponse<Order>> response) {
                btnConfirmOrder.setEnabled(true);
                btnConfirmOrder.setText("MUA NGAY");

                if (response.isSuccessful() && response.body() != null) {
                    if (response.body().isSuccess()) {
                        Toast.makeText(ProductDetailActivity.this, "Đặt hàng thành công!", Toast.LENGTH_LONG).show();
                        finish(); 
                    } else {
                        String message = response.body().getMessage();
                        Toast.makeText(ProductDetailActivity.this, "Đặt hàng thất bại: " + message, Toast.LENGTH_LONG).show();
                        Log.e(TAG, "Đặt hàng thất bại: " + message);
                    }
                } else {
                    Toast.makeText(ProductDetailActivity.this, "Lỗi Server: " + response.code(), Toast.LENGTH_SHORT).show();
                    try {
                        Log.e(TAG, "Error Body: " + response.errorBody().string());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResponse<Order>> call, @NonNull Throwable t) {
                btnConfirmOrder.setEnabled(true);
                btnConfirmOrder.setText("MUA NGAY");
                Log.e(TAG, "Đặt hàng thất bại: " + t.getMessage());
                Toast.makeText(ProductDetailActivity.this, "Không thể kết nối đến máy chủ", Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/example/doan/Activities/ProductDetailActivity.kt
package com.example.doan.Activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.doan.Adapters.ReviewAdapter
import com.example.doan.Models.*
import com.example.doan.Network.RetrofitClient
import com.example.doan.R
import com.example.doan.Utils.LoadingDialog
import com.example.doan.Utils.SessionManager
import com.google.android.material.button.MaterialButton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.Locale

class ProductDetailActivity : AppCompatActivity() {

    private lateinit var ivProductImage: ImageView
    private lateinit var tvProductName: TextView
    private lateinit var tvProductPrice: TextView
    private lateinit var tvProductDescription: TextView
    private lateinit var tvCategory: TextView
    private lateinit var tvQuantity: TextView
    private lateinit var tvTotalPrice: TextView
    private lateinit var btnBack: ImageButton
    private lateinit var btnDecrease: MaterialButton
    private lateinit var btnIncrease: MaterialButton
    private lateinit var btnAddToCart: MaterialButton
    private lateinit var btnConfirmOrder: MaterialButton
    private lateinit var spinnerSize: Spinner
    private lateinit var layoutToppings: LinearLayout
    private lateinit var tvToppingLabel: TextView
    
    // Review views
    private lateinit var layoutRatingSummary: LinearLayout
    private lateinit var tvAverageRating: TextView
    private lateinit var ratingBarAverage: RatingBar
    private lateinit var tvTotalReviews: TextView
    private lateinit var tvNoReviews: TextView
    private lateinit var rvReviews: RecyclerView
    private lateinit var tvViewAllReviews: TextView
    private lateinit var reviewAdapter: ReviewAdapter
    
    private lateinit var loadingDialog: LoadingDialog

    private var productId: Int = -1
    private var product: Drink? = null
    private var quantity = 1
    private var selectedSize: DrinkSize? = null
    private val selectedToppings = mutableSetOf<DrinkTopping>()
    
    // FIX C5: Lưu reference của các Retrofit calls để cancel khi Activity destroy
    private var loadProductCall: Call<ApiResponse<List<Drink>>>? = null
    private var addToCartCall: Call<ApiResponse<Cart>>? = null
    private var loadReviewsCall: Call<ApiResponse<List<Review>>>? = null
    private var loadRatingSummaryCall: Call<ApiResponse<DrinkRatingSummary>>? = null

    companion object {
        private const val TAG = "ProductDetailActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_product_detail_new)

        loadingDialog = LoadingDialog(this)
        
        initViews()
        getIntentData()
        setupListeners()
        setupReviewsRecyclerView()
        loadProductDetails()
    }
    
    // FIX C5: Cancel tất cả pending calls khi Activity bị destroy
    override fun onDestroy() {
        super.onDestroy()
        loadProductCall?.cancel()
        addToCartCall?.cancel()
        loadReviewsCall?.cancel()
        loadRatingSummaryCall?.cancel()
        Log.d(TAG, "Cancelled pending API calls")
    }

    private fun initViews() {
        ivProductImage = findViewById(R.id.detail_product_image)
        tvProductName = findViewById(R.id.detail_product_name)
        tvProductPrice = findViewById(R.id.detail_product_price)
        tvProductDescription = findViewById(R.id.detail_product_description)
        tvCategory = findViewById(R.id.detail_product_category)
        tvQuantity = findViewById(R.id.tv_quantity)
        tvTotalPrice = findViewById(R.id.tv_total_price)

        btnBack = findViewById(R.id.btn_back)
        btnDecrease = findViewById(R.id.btn_decrease_qty)
        btnIncrease = findViewById(R.id.btn_increase_qty)
        btnAddToCart = findViewById(R.id.btn_add_to_cart)
        btnConfirmOrder = findViewById(R.id.btn_confirm_order)

        spinnerSize = findViewById(R.id.spinner_size)
        layoutToppings = findViewById(R.id.layout_toppings)
        tvToppingLabel = findViewById(R.id.tv_topping_label)
        
        // Review views
        layoutRatingSummary = findViewById(R.id.layout_rating_summary)
        tvAverageRating = findViewById(R.id.tv_average_rating)
        ratingBarAverage = findViewById(R.id.rating_bar_average)
        tvTotalReviews = findViewById(R.id.tv_total_reviews)
        tvNoReviews = findViewById(R.id.tv_no_reviews)
        rvReviews = findViewById(R.id.rv_reviews)
        tvViewAllReviews = findViewById(R.id.tv_view_all_reviews)
    }
    
    private fun setupReviewsRecyclerView() {
        reviewAdapter = ReviewAdapter()
        rvReviews.layoutManager = LinearLayoutManager(this)
        rvReviews.adapter = reviewAdapter
        rvReviews.isNestedScrollingEnabled = false
    }

    private fun getIntentData() {
        productId = intent.getIntExtra("PRODUCT_ID", -1)
        if (productId == -1) {
            Toast.makeText(this, "Không tìm thấy sản phẩm", Toast.LENGTH_SHORT).show()
            finish()
            return
        }

        tvProductName.text = intent.getStringExtra("PRODUCT_NAME")
        tvCategory.text = intent.getStringExtra("CATEGORY_NAME")
        tvProductDescription.text = intent.getStringExtra("PRODUCT_DESC")

        val price = intent.getDoubleExtra("PRODUCT_PRICE", 0.0)
        tvProductPrice.text = String.format(Locale.getDefault(), "%,.0f VNĐ", price)
        tvTotalPrice.text = String.format(Locale.getDefault(), "%,.0f VNĐ", price)

        val imageUrl = intent.getStringExtra("PRODUCT_IMAGE")
        Glide.with(this)
            .load(imageUrl)
            .placeholder(R.drawable.ic_image_placeholder)
            .into(ivProductImage)
    }

    private fun loadProductDetails() {
        loadingDialog.show("Đang tải sản phẩm...")
        
        // FIX C5: Lưu reference để có thể cancel
        loadProductCall = RetrofitClient.getInstance(this).apiService.getDrinks()
        loadProductCall?.enqueue(object : Callback<ApiResponse<List<Drink>>> {
            override fun onResponse(call: Call<ApiResponse<List<Drink>>>, response: Response<ApiResponse<List<Drink>>>) {
                loadingDialog.dismiss()
                
                // FIX C5: Kiểm tra Activity còn tồn tại không
                if (isFinishing || isDestroyed) {
                    Log.d(TAG, "Activity destroyed, ignoring response")
                    return
                }
                
                if (response.isSuccessful && response.body()?.success == true) {
                    val drinks = response.body()?.data ?: emptyList()
                    product = drinks.find { it.id == productId }

                    product?.let {
                        displayProductFullDetails(it)
                        loadReviews(productId.toLong())
                    } ?: run {
                        // FIX C2: Hiển thị lỗi cho user khi không tìm thấy sản phẩm
                        Toast.makeText(this@ProductDetailActivity, "Không tìm thấy thông tin sản phẩm", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    // FIX C2: Xử lý lỗi response và hiển thị cho user
                    val errorMessage = try {
                        response.body()?.message ?: response.errorBody()?.string() ?: "Lỗi tải thông tin sản phẩm"
                    } catch (e: Exception) {
                        "Lỗi tải thông tin sản phẩm"
                    }
                    Toast.makeText(this@ProductDetailActivity, errorMessage, Toast.LENGTH_SHORT).show()
                    Log.e(TAG, "Error response: $errorMessage")
                }
            }

            override fun onFailure(call: Call<ApiResponse<List<Drink>>>, t: Throwable) {
                loadingDialog.dismiss()
                
                // FIX C5: Kiểm tra Activity còn tồn tại không
                if (isFinishing || isDestroyed) {
                    Log.d(TAG, "Activity destroyed, ignoring failure")
                    return
                }
                
                // FIX C2: Hiển thị lỗi mạng cho user
                if (!call.isCanceled) {
                    val errorMessage = when {
                        t is java.net.UnknownHostException -> "Không có kết nối mạng"
                        t is java.net.SocketTimeoutException -> "Kết nối quá thời gian chờ"
                        else -> "Lỗi kết nối: ${t.message}"
                    }
                    Toast.makeText(this@ProductDetailActivity, errorMessage, Toast.LENGTH_SHORT).show()
                    Log.e(TAG, "Error loading product details", t)
                }
            }
        })
    }

    private fun displayProductFullDetails(prod: Drink) {
        val sizes = prod.sizes
        if (!sizes.isNullOrEmpty()) {
            val sizeNames = sizes.map {
                if (it.extraPrice > 0)
                    "${it.sizeName} (+${String.format("%,.0f", it.extraPrice)})"
                else it.sizeName
            }
            val sizeAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, sizeNames)
            spinnerSize.adapter = sizeAdapter

            spinnerSize.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    selectedSize = sizes[position]
                    updateTotalPrice()
                }

                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }

            selectedSize = sizes[0]
        } else {
            spinnerSize.visibility = View.GONE
        }

        val toppings = prod.toppings
        if (!toppings.isNullOrEmpty()) {
            tvToppingLabel.visibility = View.VISIBLE
            layoutToppings.visibility = View.VISIBLE
            layoutToppings.removeAllViews()

            toppings.filter { it.isActive }.forEach { topping ->
                val checkBox = CheckBox(this)
                checkBox.text = "${topping.toppingName} (+${String.format("%,.0f", topping.price)})"

                checkBox.setOnCheckedChangeListener { _, isChecked ->
                    if (isChecked) selectedToppings.add(topping)
                    else selectedToppings.remove(topping)
                    updateTotalPrice()
                }

                layoutToppings.addView(checkBox)
            }
        } else {
            tvToppingLabel.visibility = View.GONE
            layoutToppings.visibility = View.GONE
        }

        updateTotalPrice()
    }
    
    private fun loadReviews(drinkId: Long) {
        // Load rating summary
        loadRatingSummaryCall = RetrofitClient.getInstance(this).apiService.getDrinkRatingSummary(drinkId)
        loadRatingSummaryCall?.enqueue(object : Callback<ApiResponse<DrinkRatingSummary>> {
            override fun onResponse(call: Call<ApiResponse<DrinkRatingSummary>>, response: Response<ApiResponse<DrinkRatingSummary>>) {
                if (isFinishing || isDestroyed) return
                
                if (response.isSuccessful && response.body()?.success == true) {
                    val summary = response.body()?.data
                    if (summary != null && summary.totalReviews > 0) {
                        layoutRatingSummary.visibility = View.VISIBLE
                        tvNoReviews.visibility = View.GONE
                        
                        tvAverageRating.text = String.format("%.1f", summary.averageRating)
                        ratingBarAverage.rating = summary.averageRating.toFloat()
                        tvTotalReviews.text = "(${summary.totalReviews} đánh giá)"
                    } else {
                        layoutRatingSummary.visibility = View.GONE
                        tvNoReviews.visibility = View.VISIBLE
                    }
                }
            }
            
            override fun onFailure(call: Call<ApiResponse<DrinkRatingSummary>>, t: Throwable) {
                if (!call.isCanceled) {
                    Log.e(TAG, "Failed to load rating summary", t)
                }
            }
        })
        
        // Load reviews list
        loadReviewsCall = RetrofitClient.getInstance(this).apiService.getReviewsByDrink(drinkId)
        loadReviewsCall?.enqueue(object : Callback<ApiResponse<List<Review>>> {
            override fun onResponse(call: Call<ApiResponse<List<Review>>>, response: Response<ApiResponse<List<Review>>>) {
                if (isFinishing || isDestroyed) return
                
                if (response.isSuccessful && response.body()?.success == true) {
                    val reviews = response.body()?.data ?: emptyList()
                    if (reviews.isNotEmpty()) {
                        rvReviews.visibility = View.VISIBLE
                        tvNoReviews.visibility = View.GONE
                        // Chỉ hiển thị tối đa 3 đánh giá gần nhất
                        val displayReviews = reviews.take(3)
                        reviewAdapter.updateReviews(displayReviews)
                        
                        if (reviews.size > 3) {
                            tvViewAllReviews.visibility = View.VISIBLE
                            tvViewAllReviews.text = "Xem tất cả (${reviews.size})"
                        }
                    } else {
                        rvReviews.visibility = View.GONE
                    }
                }
            }
            
            override fun onFailure(call: Call<ApiResponse<List<Review>>>, t: Throwable) {
                if (!call.isCanceled) {
                    Log.e(TAG, "Failed to load reviews", t)
                }
            }
        })
    }

    private fun setupListeners() {
        btnBack.setOnClickListener { finish() }

        btnDecrease.setOnClickListener {
            if (quantity > 1) {
                quantity--
                tvQuantity.text = quantity.toString()
                updateTotalPrice()
            }
        }

        btnIncrease.setOnClickListener {
            quantity++
            tvQuantity.text = quantity.toString()
            updateTotalPrice()
        }

        btnAddToCart.setOnClickListener { addToCart(false) }
        btnConfirmOrder.setOnClickListener { addToCart(true) }
    }

    private fun updateTotalPrice() {
        val basePrice = product?.basePrice ?: intent.getDoubleExtra("PRODUCT_PRICE", 0.0)
        val sizeExtra = selectedSize?.extraPrice ?: 0.0
        val toppingTotal = selectedToppings.sumOf { it.price }

        val total = (basePrice + sizeExtra + toppingTotal) * quantity
        tvTotalPrice.text = String.format(Locale.getDefault(), "%,.0f VNĐ", total)
    }

    private fun addToCart(goToCart: Boolean) {
        val session = SessionManager(this)
        if (!session.isLoggedIn()) {
            Toast.makeText(this, "Vui lòng đăng nhập để đặt hàng", Toast.LENGTH_SHORT).show()
            startActivity(Intent(this, LoginActivity::class.java))
            return
        }

        loadingDialog.show("Đang thêm vào giỏ hàng...")

        val request = AddToCartRequest(
            drinkId = productId.toLong(),
            sizeId = selectedSize?.id?.toLong() ?: 0L,
            quantity = quantity,
            toppingIds = selectedToppings.map { it.id.toLong() },
            note = ""
        )

        // FIX C5: Lưu reference để có thể cancel
        // Sử dụng API mới - không cần truyền userId (lấy từ JWT token)
        addToCartCall = RetrofitClient.getInstance(this).apiService.addToCart(request)
        addToCartCall?.enqueue(object : Callback<ApiResponse<Cart>> {
            override fun onResponse(call: Call<ApiResponse<Cart>>, response: Response<ApiResponse<Cart>>) {
                loadingDialog.dismiss()
                
                // FIX C5: Kiểm tra Activity còn tồn tại không
                if (isFinishing || isDestroyed) return
                
                if (response.isSuccessful && response.body()?.success == true) {
                    if (goToCart) {
                        // Lấy orderType từ SharedPreferences
                        val prefs = getSharedPreferences("UTETeaPrefs", MODE_PRIVATE)
                        val orderType = prefs.getString("orderType", "pickup") ?: "pickup"
                        
                        startActivity(Intent(this@ProductDetailActivity, CartActivity::class.java).apply {
                            putExtra("orderType", orderType)
                        })
                        finish()
                    } else {
                        showSuccessDialog()
                    }
                } else {
                    // FIX C2: Xử lý lỗi response tốt hơn
                    val errorMessage = try {
                        response.body()?.message ?: response.errorBody()?.string() ?: "Lỗi thêm giỏ hàng"
                    } catch (e: Exception) {
                        "Lỗi thêm giỏ hàng"
                    }
                    Toast.makeText(this@ProductDetailActivity, errorMessage, Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<ApiResponse<Cart>>, t: Throwable) {
                loadingDialog.dismiss()
                
                // FIX C5: Kiểm tra Activity còn tồn tại không
                if (isFinishing || isDestroyed) return
                
                // FIX C2: Hiển thị lỗi mạng cho user
                if (!call.isCanceled) {
                    val errorMessage = when {
                        t is java.net.UnknownHostException -> "Không có kết nối mạng"
                        t is java.net.SocketTimeoutException -> "Kết nối quá thời gian chờ"
                        else -> "Lỗi kết nối: ${t.message}"
                    }
                    Toast.makeText(this@ProductDetailActivity, errorMessage, Toast.LENGTH_SHORT).show()
                }
            }
        })
    }

    private fun showSuccessDialog() {
        AlertDialog.Builder(this)
            .setTitle("Thành công")
            .setMessage("Đã thêm món vào giỏ hàng")
            .setPositiveButton("Đến giỏ hàng") { _, _ ->
                // Lấy orderType từ SharedPreferences
                val prefs = getSharedPreferences("UTETeaPrefs", MODE_PRIVATE)
                val orderType = prefs.getString("orderType", "pickup") ?: "pickup"
                
                startActivity(Intent(this, CartActivity::class.java).apply {
                    putExtra("orderType", orderType)
                })
                finish()
            }
            .setNegativeButton("Tiếp tục mua") { d, _ -> d.dismiss() }
            .show()
    }
}
